virtio-devices/src/vhost_user/vu_common_ctrl.rs

```diff
@@ -108,13 +112,35 @@ pub fn setup_vhost_user<S: VhostUserMasterReqHandler>(
     virtio_interrupt: &Arc<dyn VirtioInterrupt>,
     acked_features: u64,
     slave_req_handler: &Option<MasterReqHandler<S>>,
+    inflight: Option<&mut Inflight>,
 ) -> Result<()> {
     vu.set_features(acked_features)
         .map_err(Error::VhostUserSetFeatures)?;
 
     // Let's first provide the memory table to the backend.
     update_mem_table(vu, mem)?;
 
+    // Setup for inflight I/O tracking shm log area.
+    if let Some(inflight) = inflight {
+        if inflight.fd.is_none() {
+            let inflight_req_info = VhostUserInflight {
+                mmap_size: 0,
+                mmap_offset: 0,
+                num_queues: queues.len() as u16,
+                queue_size: queues[0].actual_size(),
+            };
+            let (info, fd) = vu
+                .get_inflight_fd(&inflight_req_info)
+                .map_err(Error::VhostUserGetInflight)?;
+            inflight.info = info;
+            inflight.fd = Some(fd);
+        }
+        if let Some(inflight_fd) = inflight.fd.as_ref() {
+            vu.set_inflight_fd(&inflight.info, inflight_fd.as_raw_fd())
+                .map_err(Error::VhostUserSetInflight)?;
+        }
```

Comment:
One more thing, for the case where we call into `GET_INFLIGHT_FD`, we must `mmap()` the shared buffer into Cloud Hypervisor virtual address space. This will prevent the shared buffer from being freed when the backend crashes since Cloud Hypervisor will continue holding onto it. Here is the reference to QEMU's code: https://github.com/qemu/qemu/blob/7fe7fae8b48e3f9c647fd685e5155ebc8e6fb84d/hw/virtio/vhost-user.c#L2348-L2349

Reply:
> I think we really need an `if`/`else` statement here, unless I'm missing something?

If it is the first connection, `inflight.fd` would be `None`, we need first `get_inflight_fd` and then `set_inflight_fd`. If it is a reconnection, then we should only do `set_inflight_fd`. So it is not an `if`/`else` statement.

Reply:
>If it is the first connection, inflight.fd would be None, we need first get_inflight_fd and then set_inflight_fd. If it is a reconnection, then we should only do set_inflight_fd. So it is not an if/else statement.

This is not very logical to me, but I can see how QEMU does it in [here](https://github.com/qemu/qemu/blob/master/hw/block/vhost-user-blk.c#L148-L160) and QEMU is the reference for vhost-user implementation.

Reply:
>I don't think we will lose the content, as the fd is transferred to cloud-hypervisor process. And there is no problem when I am testing this PR with QEMU vhost-user-blk daemon. Actually, QEMU does not make use the mmap'd inflight->addr yet, the inflight->addr are used in some helper functions (https://github.com/qemu/qemu/blob/d874bc081600528f0400977460b4f98f21e156a1/include/hw/virtio/vhost.h#L142) that currently never called by other components. These helper functions seem to be useful if we are going to implement live-migration for the inflight area. So maybe we can implement them in later commits?

Oh yes you're totally right :)
Let's not worry about this for now.